src/main/java/org/spongepowered/api/resource/ResourceManager.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.resource;
+
+import java.util.List;
+import java.util.Optional;
+
+public interface ResourceManager {
```

Comment:
Something like this: https://gist.github.com/Cybermaxke/24c015876742730b9b1407b96e463506
There are more methods present in the current API but this will give a basic idea. In the Server/Client class are there two methods, one for the system asset manager and one that supports resource packs and custom child asset managers.

You can also look at a implementation I did a while ago: https://github.com/LanternPowered/LanternServer/tree/master/src/main/java/org/lanternpowered/server/asset
Although it can be cleaned up, since I created a custom Asset interface to add the id (resource location). And I like your idea about having an event to check when a asset manager is reloaded.